### PR TITLE
Optimize ERC721 (depends on #1030)

### PR DESCRIPTION
**Pull request by abandeali1**
This PR depends on #1030 (happy to drop that commit if it's an issue).

# 🚀 Description

There are a _ton_ of redundant checks and state reads in the current ERC721 implementations. I'm assuming that a lot of this is because everything is broken up into multiple function calls. Even though they read from a lot of the same state, the optimizer has trouble combining the SLOADs because they are occurring in different scopes. 

By flattening all of the function calls in `transferFrom`, quick tests show that we save a whopping **~8000 gas** per transfer (this is in addition to the ~2000 gas saved by dropping the modifiers in #1030). 

IMO, flattening the functions also makes actual token contracts more readable, as we are able to drop a lot of unused functions. The downside is that there is now a lot of redundant code in the mock contracts, but I think the tradeoff is massively worth it.

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

File: contracts/token/ERC721/ERC721Token.sol [hunk 1]
```diff
@@ -106,6 +106,35 @@ contract ERC721Token is SupportsInterfaceWithLookup, ERC721BasicToken, ERC721 {
...
+    public
+  {
+    super.transferFrom(_from, _to, _tokenId);
+
+    uint256 removeTokenIndex = ownedTokensIndex[_tokenId];
+    uint256 lastTokenIndex = ownedTokens[_from].length.sub(1);
+    uint256 lastToken = ownedTokens[_from][lastTokenIndex];
+
+    ownedTokens[_from][removeTokenIndex] = lastToken;
+    ownedTokens[_from][lastTokenIndex] = 0;
```
Comment (johnhforrest) [inline contracts/token/ERC721/ERC721Token.sol]: As per https://github.com/OpenZeppelin/openzeppelin-solidity/issues/977 this line should be safe to remove and save a bit more gas.

**Comment by johnhforrest** [comment]
Huge fan of this PR. Really helps from a readability perspective and reduces the coupling on the burn/mint logic. Great work.

**Comment by mg6maciej** [comment]
Why did you remove `_mint` (and `_burn`)? It's such a useful function that is exposed in derived contracts after applying some verification logic.

**Comment by abandeali1** [comment]
@mg6maciej I didn't want to leave in any unused functions just for the sake of using them in derived contracts. 

If you're asking why they were removed in the first place, see the PR description. The gist of it is that they are horribly inefficient.

**Comment by mg6maciej** [comment]
I wouldn't want to force users to understand the internals of ERC721 implementation to use it. If you try to derive from this class, you will quickly see, why it's better to have `_mint` there.
Other implementations [here](https://github.com/mg6maciej/erc721-implementation/blob/master/contracts/ERC721.sol#L146), [here](https://github.com/0xcert/ethereum-erc721/blob/master/contracts/tokens/NFToken.sol#L368) provide it as well. If you search github for OpenZeppelin imports, you will see people expect it to be there: [here](https://github.com/Userfeeds/robohash-token/blob/only_token/contracts/RoboHashToken.sol#L13), [here](https://github.com/ultiplex/SkilleX/blob/master/contracts/Skillex.sol#L24), [here](https://github.com/levelkdev/nft-exchange/blob/1fc3506431740f1c8f26a0bafcbcf75d08adda02/contracts/ExampleNFT.sol#L9), [here](https://github.com/cipherzzz/erc721/blob/516a2ff33d9924ca47c6f433fd11f87af6119218/contracts/MyERC721.sol#L20), [here](https://github.com/fMercury/MetaCoin-tp/blob/0539f105a795fa8a95b012837a3eaa02676c0824/Patricio-Garcia-Zitarosa/ERC-721/contracts/HubbingERC721.sol#L10), [here](https://github.com/Emiller88/Crypto-Adventures/blob/97f8a07e189e90100197b557e761abeb4d00525b/contracts/createItem.sol#L9), [here](https://github.com/mattlockyer/composables-998/blob/22fe85561b3f4d39deaf79c73b93d6ce7061765a/contracts/SampleNFT.sol#L20), [here](https://github.com/beyondns/nft/blob/a1009e34fc49a16551aa5ad804a42d60b3c73ae5/contracts/erc721mintabletoken.sol#L7)...

**Comment by abandeali1** [comment]
Why not just split these functions out into a separate contract? I think they are useful, but I just don't think they should be used internally within `transferFrom`. And it feels wasteful to increase the cost of deployment for anyone not using these functions.

Happy to add this if it is a blocker.

**Comment by nventuro** [comment]
Hello @abandeali1, thank you for your contribution! 

I dug a bit deeper into what was causing the increased gas cost (since it seemed too high for it to come from just a couple extra `SLOAD`s, and found that the reason lies behind the `_removeTokenFrom` and `_addTokenTo` calls: the compiler isn't smart enough to realize there's no need to delete it from the original user because it can simply be re-assigned.

This can be more easily fixed by adding a `private` function to handle the special case of transferring (since `add` and `remove` need to continue existing to support `mint` and `burn`): I've tested this and it achieves similar gas improvements (there is a total gas cost of less than 1% between this and your approach).

Because that's a less disruptive change (we do want `_mint` and `_burn`), and this PR got outdated after the 2.0 release (sorry about the delay!), I'm closing this in favor of https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1539. Once again, thanks to @abandeali1 for pushing for this, and to everyone who participated in the discussion and validated this idea!

**Pull request closed without merging**